### PR TITLE
feat/New formula for calculations and estimated gas fee

### PR DESCRIPTION
**Pull request by arlindlinum**
IMPORTANT NOTE:

We added a new env which need to be added on your hosting provider also on code for locally on .env.local

NEXT_PUBLIC_MAINNET_CLIENT="https://eth-mainnet.g.alchemy.com/v2/yb_t_jZK-Ql8bD1WYT-v_6Lzn0btd3xh"

### CODE REVIEW

File: components/token-container.tsx [hunk 1]
```diff
@@ -18,11 +21,44 @@ type TokenContainerProps = FlexProps & {
...
+  calculatePriceAndToken?: () => void;
 };
 
 export const TokenContainer = (props: TokenContainerProps) => {
   const ethCard = props.symbol === "ETH";
 
+  const maxValueWithGas = async () => {
+    if (!props.maxBalance || parseFloat(props.maxBalance) === 0) return "0";
+
+    const defaultGasEstimate = parseEther("0.0045");
```
Comment (jiggyjo11) [inline components/token-container.tsx]: would be great creating a const for all of that and and import it from a central place... (for the numbers)

File: components/token-container.tsx [hunk 2]
```diff
@@ -18,11 +21,44 @@ type TokenContainerProps = FlexProps & {
...
+    const valueAsBigNumber = parseEther(props.maxBalance);
```
Comment (jiggyjo11) [inline components/token-container.tsx]: you can destructure props
const {maxBalance, otherProp} = maxBalance;

File: components/token-container.tsx [hunk 3]
```diff
@@ -18,11 +21,44 @@ type TokenContainerProps = FlexProps & {
...
+    const maxGasUsage = 100000; // Max gas usage seen on Etherscan was ~85k
```
Comment (jiggyjo11) [inline components/token-container.tsx]: import these from central place

File: components/token-container.tsx [hunk 4]
```diff
@@ -18,11 +21,44 @@ type TokenContainerProps = FlexProps & {
...
+
+    try {
+      const feeData = await estimateFeesPerGas(mainnetClient);
+
+      if (feeData) {
```
Comment (jiggyjo11) [inline components/token-container.tsx]: no need to wrap the two if statements, you can check the conditions like this 
`if ( feeData && feeData.maxFeePerGas)`

File: components/token-container.tsx [hunk 5]
```diff
@@ -18,11 +21,44 @@ type TokenContainerProps = FlexProps & {
...
+            if (gasCost < defaultGasEstimate) {
+              gasCost = defaultGasEstimate;
+            }
+            const valMinusGas = valueAsBigNumber - gasCost;
+            if (valMinusGas <= 0n) return "0";
+            return formatEther(valMinusGas);
+          }
+        }
+      }
+    } catch (error) {}
```
Comment (jiggyjo11) [inline components/token-container.tsx]: you shoud log error so that we see it in vercel

File: components/token-container.tsx [hunk 6]
```diff
@@ -56,9 +92,18 @@ export const TokenContainer = (props: TokenContainerProps) => {
               display={"flex"}
               h={"fit-content"}
               p={"2px 6px"}
-              onClick={() => {
-                if (props.maxBalance) {
-                  props.setAmount(props.maxBalance);
+              onClick={async () => {
```
Comment (jiggyjo11) [inline components/token-container.tsx]: dont pass entire functions like that to the onClick prop, create a handler function above the return statement that renders jsx and simply reference it in onClick. This is easier to read and much better in terms of performance

File: components/windows/swap-widget.tsx [hunk 7]
```diff
@@ -277,6 +279,14 @@ export const SwapWidget = () => {
                         image="/windows/swap/ETH.svg"
                         maxBalance={formattedEthBalance}
                         setFocused={setFocused}
+                        calculatePriceAndToken={() =>
```
Comment (jiggyjo11) [inline components/windows/swap-widget.tsx]: much better to not use arrow functions in props but rather creating handler functions above jsx.. this makes the whole component much easier to read and uses less compute e.g making app more performant

File: components/token-container.tsx (hunk 1 above)
Comment (arlindlinum) [reply components/token-container.tsx]: Done

File: components/token-container.tsx (hunk 2 above)
Comment (arlindlinum) [reply components/token-container.tsx]: Done

File: components/token-container.tsx (hunk 3 above)
Comment (arlindlinum) [reply components/token-container.tsx]: Done

File: components/token-container.tsx (hunk 4 above)
Comment (arlindlinum) [reply components/token-container.tsx]: Done

File: components/token-container.tsx (hunk 5 above)
Comment (arlindlinum) [reply components/token-container.tsx]: Done

File: components/token-container.tsx (hunk 6 above)
Comment (arlindlinum) [reply components/token-container.tsx]: Done

File: components/windows/swap-widget.tsx (hunk 7 above)
Comment (arlindlinum) [reply components/windows/swap-widget.tsx]: Done

**Pull request merged**